kotlin-web-spring-boot-3/src/main/kotlin/com/vndevteam/kotlinwebspringboot3/application/logging/CustomResponseBody.kt

```diff
@@ -0,0 +1,83 @@
+import jakarta.servlet.http.HttpServletRequest
+import jakarta.servlet.http.HttpServletResponse
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.core.MethodParameter
+import org.springframework.http.MediaType
+import org.springframework.http.converter.HttpMessageConverter
+import org.springframework.http.server.ServerHttpRequest
+import org.springframework.http.server.ServerHttpResponse
+import org.springframework.http.server.ServletServerHttpRequest
+import org.springframework.http.server.ServletServerHttpResponse
+import org.springframework.web.bind.annotation.ControllerAdvice
+import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice
+import java.sql.Timestamp
+
+@ControllerAdvice
+class CustomResponseBody : ResponseBodyAdvice<Any> {
+    companion object {
+        val log: Logger = LoggerFactory.getLogger(CustomResponseBody::class.java)
+    }
+    
+    override fun supports(returnType: MethodParameter, converterType: Class<out HttpMessageConverter<*>>): Boolean {
+        return true
+    }
+
+    override fun beforeBodyWrite(
+        body: Any?,
+        returnType: MethodParameter,
+        selectedContentType: MediaType,
+        selectedConverterType: Class<out HttpMessageConverter<*>>,
+        request: ServerHttpRequest,
+        response: ServerHttpResponse
+    ): Any? {
+        if (request is ServletServerHttpRequest && response is ServletServerHttpResponse) {
+            logResponse(
+                request.servletRequest,
+                response.servletResponse, body
+            )
+        }
+
+        return body
+    }
+
+    private fun logResponse(
+        httpServletRequest: HttpServletRequest,
+        httpServletResponse: HttpServletResponse,
+        body: Any?
+    ) {
+        val requestId = httpServletRequest.getAttribute(RequestFilter.REQUEST_ID)
+        if (requestId != null) {
+            val startTime = httpServletRequest.getAttribute(RequestFilter.START_TIME) as Long
+            val endTime = System.currentTimeMillis()
+            val data = StringBuilder()
+            data.append(
+                "Request URL: ${httpServletRequest.requestURL} - Take Time: ${endTime - startTime} millisecond (start: ${
+                    Timestamp(
+                        startTime
+                    )
+                }, end: ${Timestamp(endTime)})."
+            )
+
+            var bodyString: String? = null
+            try {
+                bodyString = MapperUtils.objectMapperForKotlin().writeValueAsString(body)
+            } catch (e: Exception) {
+                log.warn("Parse json fail: {}", e.message)
+            }
+
+            if (httpServletResponse.status.toString().startsWith("2")) {
+                log.info(data.toString())
+            } else if (httpServletResponse.status.toString().startsWith("4")) {
+                data.append("Response body: ").append(bodyString).append(". ")
```

Comment:
@haiks-2382 Phần log response H tham khảo bài này để log ra đầy đủ thông tin hơn: https://medium.com/@ankithahjpgowda/log-request-and-responses-of-rest-apis-in-springboot-c13f9bc7903f